src/joystick/SDL_gamepad.c

```diff
@@ -2955,48 +2956,49 @@ int SDL_SendGamepadEffect(SDL_Gamepad *gamepad, const void *data, int size)
     return SDL_SendJoystickEffect(joystick, data, size);
 }
 
-void SDL_CloseGamepad(SDL_Gamepad *gamepad)
+int SDL_CloseGamepad(SDL_Gamepad *gamepad)
 {
     SDL_Gamepad *gamepadlist, *gamepadlistprev;
 
     SDL_LockJoysticks();
+    {
+        CHECK_GAMEPAD_MAGIC(gamepad, -1);
 
-    if (gamepad == NULL || gamepad->magic != &gamepad_magic) {
-        SDL_UnlockJoysticks();
-        return;
-    }
-
-    /* First decrement ref count */
-    if (--gamepad->ref_count > 0) {
-        SDL_UnlockJoysticks();
-        return;
-    }
+        /* First decrement ref count */
+        if (--gamepad->ref_count > 0) {
+            SDL_UnlockJoysticks();
+            return 0;
+        }
 
-    SDL_CloseJoystick(gamepad->joystick);
+        if (SDL_CloseJoystick(gamepad->joystick) != 0) {
+            /* SDL_SetError() called within SDL_CloseJoystick() */
+            return -1;
+        }
```

Comment:
I'm unsure about the above __if__ block. Do all gamepads have joysticks. My quick skim of the code seemed to point in that direction but I'm not 100%. Some assurance would be nice.

Reply:
They all do, yes, but a failure here shouldn't prevent the rest of the cleanup. See my comment above about close/destroy functions in general.